support OpenSSL 1.1

Fedora Rawhide has dumped OSSL 1.1 in their repository, the build log down below is the result of trying to build against it

```
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp: In function ‘DH* my_get_dh(int)’:
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:212:4: error: invalid use of incomplete type ‘DH {aka struct dh_st}’
  dh->p = BN_bin2bn(p, sp, nullptr);
    ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:104:16: note: forward declaration of ‘DH {aka struct dh_st}’
 typedef struct dh_st DH;
                ^~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:213:4: error: invalid use of incomplete type ‘DH {aka struct dh_st}’
  dh->g = BN_bin2bn(g, sg, nullptr);
    ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:104:16: note: forward declaration of ‘DH {aka struct dh_st}’
 typedef struct dh_st DH;
                ^~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:214:8: error: invalid use of incomplete type ‘DH {aka struct dh_st}’
  if((dh->p == nullptr) || (dh->g == nullptr))
        ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:104:16: note: forward declaration of ‘DH {aka struct dh_st}’
 typedef struct dh_st DH;
                ^~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:214:30: error: invalid use of incomplete type ‘DH {aka struct dh_st}’
  if((dh->p == nullptr) || (dh->g == nullptr))
                              ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:104:16: note: forward declaration of ‘DH {aka struct dh_st}’
 typedef struct dh_st DH;
                ^~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp: In member function ‘int KviSSLCertificate::fingerprintDigestId()’:
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:698:31: error: invalid use of incomplete type ‘X509 {aka struct x509_st}’
  int NID = OBJ_obj2nid(m_pX509->sig_alg->algorithm);
                               ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:119:16: note: forward declaration of ‘X509 {aka struct x509_st}’
 typedef struct x509_st X509;
                ^~~~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:713:15: error: invalid use of incomplete type ‘const EVP_MD {aka const struct evp_md_st}’
  return mdType->type;
               ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:91:16: note: forward declaration of ‘EVP_MD {aka struct evp_md_st}’
 typedef struct evp_md_st EVP_MD;
                ^~~~~~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp: In member function ‘void KviSSLCertificate::extractPubKeyInfo()’:
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:832:22: error: invalid use of incomplete type ‘EVP_PKEY {aka struct evp_pkey_st}’
   m_szPubKeyType = (p->type == NID_undef) ? __tr("Unknown") : OBJ_nid2ln(p->type);
                      ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:93:16: note: forward declaration of ‘EVP_PKEY {aka struct evp_pkey_st}’
 typedef struct evp_pkey_st EVP_PKEY;
                ^~~~~~~~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:832:75: error: invalid use of incomplete type ‘EVP_PKEY {aka struct evp_pkey_st}’
 _szPubKeyType = (p->type == NID_undef) ? __tr("Unknown") : OBJ_nid2ln(p->type);
                                                                        ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:93:16: note: forward declaration of ‘EVP_PKEY {aka struct evp_pkey_st}’
 typedef struct evp_pkey_st EVP_PKEY;
                ^~~~~~~~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp: In member function ‘void KviSSLCertificate::extractSignature()’:
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:855:29: error: invalid use of incomplete type ‘X509 {aka struct x509_st}’
  int i = OBJ_obj2nid(m_pX509->sig_alg->algorithm);
                             ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:119:16: note: forward declaration of ‘X509 {aka struct x509_st}’
 typedef struct x509_st X509;
                ^~~~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:860:24: error: invalid use of incomplete type ‘X509 {aka struct x509_st}’
  for(i = 0; i < m_pX509->signature->length; i++)
                        ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:119:16: note: forward declaration of ‘X509 {aka struct x509_st}’
 typedef struct x509_st X509;
                ^~~~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:864:50: error: invalid use of incomplete type ‘X509 {aka struct x509_st}’
   m_szSignatureContents.append(hexdigits[(m_pX509->signature->data[i] & 0xf0) >> 4]);
                                                  ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:119:16: note: forward declaration of ‘X509 {aka struct x509_st}’
 typedef struct x509_st X509;
                ^~~~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:865:50: error: invalid use of incomplete type ‘X509 {aka struct x509_st}’
   m_szSignatureContents.append(hexdigits[(m_pX509->signature->data[i] & 0x0f)]);
                                                  ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:119:16: note: forward declaration of ‘X509 {aka struct x509_st}’
 typedef struct x509_st X509;
                ^~~~~~~
gmake[2]: *** [src/kvilib/CMakeFiles/kvilib.dir/build.make:1599: src/kvilib/CMakeFiles/kvilib.dir/net/KviSSL.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:4125: src/kvilib/CMakeFiles/kvilib.dir/all] Error 2
gmake: *** [Makefile:128: all] Error 2
```

--- a/src/kvilib/net/KviSSL.cpp
+++ b/src/kvilib/net/KviSSL.cpp
@@ -168,6 +168,7 @@ static DH * my_get_dh(int keylength)
 	unsigned char * g = nullptr;
 	int sp = 0;
 	int sg = 0;
+	BIGNUM *bp, *bg;
 	switch(keylength)
 	{
 		case 512:
@@ -209,13 +210,21 @@ static DH * my_get_dh(int keylength)
 	dh = DH_new();
 	if(!dh)
 		return nullptr;
-	dh->p = BN_bin2bn(p, sp, nullptr);
-	dh->g = BN_bin2bn(g, sg, nullptr);
-	if((dh->p == nullptr) || (dh->g == nullptr))
+	bp = BN_bin2bn(p, sp, nullptr);
+	bg = BN_bin2bn(g, sg, nullptr);
+	if((p == nullptr) || (g == nullptr))
 	{
+		BN_free(bp);
+		BN_free(bg);
 		DH_free(dh);
 		return nullptr;
 	}
+#if OPENSSL_VERSION_NUMBER >= 0x10100005L
+	DH_set0_pqg(dh, bp, nullptr, bg);
+#else
+	dh->p = bp;
+	dh->g = bg;
+#endif
 	return dh;
 }
 
@@ -695,7 +704,14 @@ int KviSSLCertificate::fingerprintDigestId()
 	if(!m_pX509)
 		return -1;
 
-	int NID = OBJ_obj2nid(m_pX509->sig_alg->algorithm);
+	const X509_ALGOR *alg;
+#if OPENSSL_VERSION_NUMBER >= 0x10100005L
+	X509_get0_signature(nullptr, &alg, m_pX509);
+#else
+	alg = m_pX509->sig_alg;
+#endif
+
+	int NID = OBJ_obj2nid(alg->algorithm);
 	if(NID == NID_undef)
 	{
 		return 0; // unknown digest function: it means the signature can't be verified: the certificate can't be trusted
@@ -710,7 +726,7 @@ int KviSSLCertificate::fingerprintDigestId()
 		return 0; // Unknown digest
 	}
 
-	return mdType->type;
+	return EVP_MD_type(mdType);
 }
 
 const char * KviSSLCertificate::fingerprintDigestStr()
@@ -828,8 +844,14 @@ void KviSSLCertificate::extractPubKeyInfo()
 	EVP_PKEY * p = X509_get_pubkey(m_pX509);
 	if(p)
 	{
+		int type;
+#if OPENSSL_VERSION_NUMBER >= 0x10100005L
+		type = EVP_PKEY_base_id(p);
+#else
+		type = EVP_PKEY_type(p->type);
+#endif
 		m_iPubKeyBits = EVP_PKEY_bits(p);
-		m_szPubKeyType = (p->type == NID_undef) ? __tr("Unknown") : OBJ_nid2ln(p->type);
+		m_szPubKeyType = (type == NID_undef) ? __tr("Unknown") : OBJ_nid2ln(type);
 		//		getPKeyType(p->type,m_szPubKeyType);
 	}
 	else
@@ -852,17 +874,26 @@ void KviSSLCertificate::extractSignature()
 {
 	static char hexdigits[] = { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F' };
 
-	int i = OBJ_obj2nid(m_pX509->sig_alg->algorithm);
+	const ASN1_BIT_STRING *sig;
+	const X509_ALGOR *alg;
+#if OPENSSL_VERSION_NUMBER >= 0x10100005L
+	X509_get0_signature(&sig, &alg, m_pX509);
+#else
+	sig = m_pX509->signature;
+	alg = m_pX509->sig_alg;
+#endif
+
+	int i = OBJ_obj2nid(alg->algorithm);
 	m_szSignatureType = (i == NID_undef) ? __tr("Unknown") : OBJ_nid2ln(i);
 
 	m_szSignatureContents = "";
 
-	for(i = 0; i < m_pX509->signature->length; i++)
+	for(i = 0; i < sig->length; i++)
 	{
 		if(m_szSignatureContents.hasData())
 			m_szSignatureContents.append(":");
-		m_szSignatureContents.append(hexdigits[(m_pX509->signature->data[i] & 0xf0) >> 4]);
-		m_szSignatureContents.append(hexdigits[(m_pX509->signature->data[i] & 0x0f)]);
+		m_szSignatureContents.append(hexdigits[(sig->data[i] & 0xf0) >> 4]);
+		m_szSignatureContents.append(hexdigits[(sig->data[i] & 0x0f)]);
 	}
 }
 

--- a/src/modules/file/libkvifile.cpp
+++ b/src/modules/file/libkvifile.cpp
@@ -53,6 +53,12 @@
 #endif
 
 #include <openssl/evp.h>
+
+#if OPENSSL_VERSION_NUMBER < 0x10100005L
+#define EVP_MD_CTX_new EVP_MD_CTX_create
+#define EVP_MD_CTX_free EVP_MD_CTX_destroy
+#endif
+
 #else
 // The fallback we can always use, but with very limited set of
 // functionality.
@@ -1547,7 +1553,7 @@ static bool file_kvs_fnc_digest(KviKvsModuleFunctionCall * c)
 	if(szAlgo.isEmpty())
 		szAlgo = "md5";
 
-	EVP_MD_CTX mdctx;
+	EVP_MD_CTX *mdctx;
 	const EVP_MD * pMD;
 	unsigned char ucMDValue[EVP_MAX_MD_SIZE];
 	unsigned int uMDLen, u;
@@ -1561,11 +1567,11 @@ static bool file_kvs_fnc_digest(KviKvsModuleFunctionCall * c)
 		return true;
 	}
 
-	EVP_MD_CTX_init(&mdctx);
-	EVP_DigestInit_ex(&mdctx, pMD, nullptr);
-	EVP_DigestUpdate(&mdctx, content.constData(), content.size());
-	EVP_DigestFinal_ex(&mdctx, ucMDValue, &uMDLen);
-	EVP_MD_CTX_cleanup(&mdctx);
+	mdctx = EVP_MD_CTX_new();
+	EVP_DigestInit_ex(mdctx, pMD, nullptr);
+	EVP_DigestUpdate(mdctx, content.constData(), content.size());
+	EVP_DigestFinal_ex(mdctx, ucMDValue, &uMDLen);
+	EVP_MD_CTX_free(mdctx);
 
 	for(u = 0; u < uMDLen; u++)
 	{

--- a/src/modules/fish/libkvifish.cpp
+++ b/src/modules/fish/libkvifish.cpp
@@ -81,25 +81,34 @@ static bool fish_DH1080_gen(unsigned char ** szPubKey, int * iLen)
 	if(!g_fish_dh)
 	{
 		BIGNUM * dhp = BN_new();
-		BN_init(dhp);
 		if(!BN_hex2bn(&dhp, g_fish_prime1080_hex))
 			return false;
 
 		BIGNUM * dhg = BN_new();
-		BN_init(dhg);
 		if(!BN_hex2bn(&dhg, g_fish_generator))
 			return false;
 
 		g_fish_dh = DH_new();
+#if OPENSSL_VERSION_NUMBER >= 0x10100005L
+		DH_set0_pqg(g_fish_dh, dhp, nullptr, dhg);
+#else
 		g_fish_dh->p = dhp;
 		g_fish_dh->g = dhg;
+#endif
 
 		DH_generate_key(g_fish_dh);
 	}
 
-	*iLen = BN_num_bytes(g_fish_dh->pub_key);
+	const BIGNUM* pub_key;
+#if OPENSSL_VERSION_NUMBER >= 0x10100005L
+	DH_get0_key(g_fish_dh, &pub_key, nullptr);
+#else
+	pub_key = g_fish_dh->pub_key;
+#endif
+
+	*iLen = BN_num_bytes(pub_key);
 	*szPubKey = (unsigned char *)KviMemory::allocate(*iLen);
-	BN_bn2bin(g_fish_dh->pub_key, *szPubKey);
+	BN_bn2bin(pub_key, *szPubKey);
 
 	return true;
 #else

--- a/src/modules/str/libkvistr.cpp
+++ b/src/modules/str/libkvistr.cpp
@@ -43,6 +43,12 @@
 #include "KviSSL.h"
 #include <openssl/evp.h>
 #include <openssl/pem.h>
+
+#if OPENSSL_VERSION_NUMBER < 0x10100005L
+#define EVP_MD_CTX_new EVP_MD_CTX_create
+#define EVP_MD_CTX_free EVP_MD_CTX_destroy
+#endif
+
 #else
 // The fallback we can always use, but with very limited set of
 // functionality.
@@ -1383,7 +1389,7 @@ static bool str_kvs_fnc_digest(KviKvsModuleFunctionCall * c)
 	if(szType.isEmpty())
 		szType = "md5";
 
-	EVP_MD_CTX mdctx;
+	EVP_MD_CTX *mdctx;
 	const EVP_MD * md;
 	unsigned char md_value[EVP_MAX_MD_SIZE];
 	unsigned int md_len, i;
@@ -1397,11 +1403,11 @@ static bool str_kvs_fnc_digest(KviKvsModuleFunctionCall * c)
 		return true;
 	}
 
-	EVP_MD_CTX_init(&mdctx);
-	EVP_DigestInit_ex(&mdctx, md, nullptr);
-	EVP_DigestUpdate(&mdctx, szString.toUtf8().data(), szString.toUtf8().length());
-	EVP_DigestFinal_ex(&mdctx, md_value, &md_len);
-	EVP_MD_CTX_cleanup(&mdctx);
+	mdctx = EVP_MD_CTX_new();
+	EVP_DigestInit_ex(mdctx, md, nullptr);
+	EVP_DigestUpdate(mdctx, szString.toUtf8().data(), szString.toUtf8().length());
+	EVP_DigestFinal_ex(mdctx, md_value, &md_len);
+	EVP_MD_CTX_free(mdctx);
 
 	for(i = 0; i < md_len; i++)
 	{
@@ -2315,10 +2321,11 @@ static bool str_kvs_fnc_evpSign(KviKvsModuleFunctionCall * c)
 #if defined(COMPILE_SSL_SUPPORT)
 
 	KviSSL::globalSSLInit();
-	EVP_MD_CTX md_ctx;
+	EVP_MD_CTX *md_ctx;
 	EVP_PKEY * pKey = nullptr;
 	unsigned int len = 0;
 	unsigned char * sig = nullptr;
+	int err;
 
 	if(szCert.isEmpty())
 	{
@@ -2369,9 +2376,12 @@ static bool str_kvs_fnc_evpSign(KviKvsModuleFunctionCall * c)
 	len = EVP_PKEY_size(pKey);
 	sig = (unsigned char *)KviMemory::allocate(len * sizeof(char));
 
-	EVP_SignInit(&md_ctx, EVP_sha1());
-	EVP_SignUpdate(&md_ctx, (unsigned char *)szMessage.data(), szMessage.length());
-	if(EVP_SignFinal(&md_ctx, sig, &len, pKey))
+	md_ctx = EVP_MD_CTX_new();
+	EVP_SignInit(md_ctx, EVP_sha1());
+	EVP_SignUpdate(md_ctx, (unsigned char *)szMessage.data(), szMessage.length());
+	err = EVP_SignFinal(md_ctx, sig, &len, pKey);
+	EVP_MD_CTX_free(md_ctx);
+	if(err)
 	{
 		QByteArray szSign((const char *)sig, len);
 		OPENSSL_free(sig);
@@ -2452,7 +2462,7 @@ static bool str_kvs_fnc_evpVerify(KviKvsModuleFunctionCall * c)
 	szSign = QByteArray::fromBase64(szSignB64);
 	const char * message = szMessage.data();
 
-	EVP_MD_CTX md_ctx;
+	EVP_MD_CTX* md_ctx;
 	EVP_PKEY * pKey = nullptr;
 	X509 * cert = nullptr;
 	int err = -1;
@@ -2519,10 +2529,11 @@ static bool str_kvs_fnc_evpVerify(KviKvsModuleFunctionCall * c)
 		}
 	}
 
-	EVP_VerifyInit(&md_ctx, EVP_sha1());
-	EVP_VerifyUpdate(&md_ctx, message, strlen(message));
-	err = EVP_VerifyFinal(&md_ctx, (unsigned char *)szSign.data(), szSign.size(), pKey);
-	EVP_MD_CTX_cleanup(&md_ctx);
+	md_ctx = EVP_MD_CTX_new();
+	EVP_VerifyInit(md_ctx, EVP_sha1());
+	EVP_VerifyUpdate(md_ctx, message, strlen(message));
+	err = EVP_VerifyFinal(md_ctx, (unsigned char *)szSign.data(), szSign.size(), pKey);
+	EVP_MD_CTX_free(md_ctx);
 	EVP_PKEY_free(pKey);
 	switch(err)
 	{

